src/main/java/com/avispl/symphony/dal/avdevices/encoderdecoder/haivision/x4decoder/HaivisionX4DecoderCommunicator.java

```diff
@@ -0,0 +1,610 @@
+								stats.put(audioPair + item.getName(), checkForNullData(audioPairs.get(i).getValueByAudioPairMetric(item)));
+							}
+						}
+					}
+				} else {
+					updateDecoderStatisticsFailedMonitor(failedMonitor, decoderID);
+				}
+			} else {
+				updateDecoderStatisticsFailedMonitor(failedMonitor, decoderID);
+			}
+		} catch (Exception e) {
+			updateDecoderStatisticsFailedMonitor(failedMonitor, decoderID);
+		}
+	}
+
+	/**
+	 * Update failedMonitor with Getting stream statistics error message
+	 *
+	 * @param failedMonitor list statistics property
+	 */
+	private void updateStreamStatisticsFailedMonitor(Map<String, String> failedMonitor) {
+		failedMonitor.put(MonitoringMetricGroup.STREAM_STATISTICS.getName(), DecoderConstant.GETTING_STREAM_STATS_ERR);
+	}
+
+	/**
+	 * This method is used update stream statistic from DTO
+	 *
+	 * @param stats list statistics property
+	 * @param stream pair of stream config and stats
+	 */
+	private void updateStreamStats(Map<String, String> stats, Stream stream) {
+		SRT srt = stream.getStreamStats().getSrt();
+		String streamStatisticGroup = MonitoringMetricGroup.STREAM_STATISTICS.getName() + DecoderConstant.SPACE + stream.getStreamInfo().getName() + DecoderConstant.HASH;
+
+		for (StreamMonitoringMetric item : StreamMonitoringMetric.values()) {
+			stats.put(streamStatisticGroup + item.getName(), checkForNullData(stream.getValueByStreamMonitoringMetric(item)));
+		}
+
+		if (stream.getStreamStats().getSrt() != null) {
+			for (SRTMetric item : SRTMetric.values()) {
+				stats.put(streamStatisticGroup + item.getName(), checkForNullData(srt.getValueBySRTMonitoringMetric(item)));
+			}
+		}
+	}
+
+	/**
+	 * This method is used to retrieve streams statistics by send GET request to http://{IP_Address}apis/streams
+	 *
+	 * @param stats list statistics property
+	 *
+	 * When sessionID is null, the failedMonitor is going to update
+	 * When there is no response data, the failedMonitor is going to update
+	 * When there is an exception, the failedMonitor is going to update and exception is not populated
+	 */
+	private void retrieveStreamStats(Map<String, String> stats) {
+
+		// Retrieve Adapter Properties
+		if (this.streamName != null && streamNameSet == null) {
+			streamNameSet = handleAdapterPropertiesInputFromUser(this.streamName);
+		}
+		if (this.streamStatus != null && streamStatusSet == null) {
+			streamStatusSet = handleAdapterPropertiesInputFromUser(this.streamStatus);
+		}
+		if (this.portNumber != null && portNumberSet == null) {
+			portNumberSet = handleAdapterPropertiesInputFromUser(this.portNumber);
+		}
+
+		try {
+			if (this.authenticationCookie.getSessionID() != null) {
+				JsonNode response = doGet(buildDeviceFullPath(DecoderURL.BASE_URI + DecoderURL.STREAMS), JsonNode.class);
+
+				if (!response.isEmpty()) {
+					objectMapper = new ObjectMapper();
+					StreamData streamData = objectMapper.readValue(response.toString(), StreamData.class);
```

Comment:
maybe it's better to use this DTO in a doGet on line 459 instead of doing that with an objectMapper, which is instantiated separately?

Reply:
I fixed it in retrieveStreamStats and retrieveDecoderStats, retrieveDeviceInfo, retrieveRoleBased as well